### PR TITLE
Improve perf for mem efficient grad mgmt

### PR DESCRIPTION
**Pull request by pengwa**
### Improve perf for mem efficient grad mgmt

When memory efficient gradient mangement feature is enabled, the weight retrieval PythonOp for every layers will be launched at the beginning of the forward, which would make GPU stream idle for few milliseconds. The reason is the ReversedDFS ordering cannot ALWAYS handle such input branching well, so we introduce a distantance-to-input_leaf concepts when doing the reversedDFS, which not only move the problematical PythonOp to the place where it is needed, but also those Cast ops following the weight retrieval to the place where it is needed.

Main branch: 102.19 - 26.35s = 75.84s for 260 steps(4627samples), 61.04sample/second
This PR: 100.28s - 25.10s = 75.18s for 260 steps. 61.54samples/second (+0.8% gains)

Main branch:

image

This PR:

image

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -1933,6 +1949,38 @@ void SortForwardNodesByReverseDFS(const Graph* graph,
   // Note 2: While it is also possible some nodes not contributing to the forward output nodes will be
   //   executed before YieldOp, for example, if one forward node's output is used by Shape/Size, then
   //   the Shape/Size node should be executed before YieldOp to release the memory as soon as possible.
+  InlinedVector<size_t> nodes_in_degree;
+  std::queue<const Node*> to_visit;
```
Comment (AdamLouly) [inline onnxruntime/core/graph/graph.cc:1953]: (I'm not sure about this) Since you've added the 'distance-to-input_leaf' concept now, would it be better to use a priority_queue here with the NodeCompareByMaxDistance function for comparison?

File: orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py [hunk 2]
```diff
@@ -38,7 +38,8 @@ def get_params_not_connected_to_pull_param_trigger(
 def post_processing_enable_mem_efficient_training(
     exported_model: ModelProto,
     named_params: dict[str, torch.nn.parameter.Parameter],
-) -> tuple[bool, ModelProto]:
+    device: torch.device,
+) -> tuple[bool, ModelProto, torch.Tensor]:
     """This function is used to enable zero stage3 compatibility.
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py]: Thanks, this is a typo :)

File: orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py [hunk 3]
```diff
@@ -141,75 +116,35 @@ def _get_param_pull_trigger_name(param_name: str) -> str:
...
-        {
-            "forward": forward,
-            "backward": backward,
-            "infer_shape": infer_shape,
-        },
-    )
-
-
-def _create_param_retrieval_function(trainable_named_params: dict[str, torch.nn.parameter.Parameter]):
+def _create_param_retrieval_function(
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_mem_efficient_grad_mgmt.py:134]: Sure. Let me add it.

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (pengwa) [reply onnxruntime/core/graph/graph.cc:1961]: It may work in some cases, while still possible we get a bad ordering for example, if we have few input leaf nodes:
Transpose(weight_layer_11), Transpose(weight_layer_10), ..., Transpose(weight_layer_1), Transpose(weight_layer_0), ATen(embedding). 

If we accidentally start with layer_11's Transpose, then its children; then layer_10's Transpose, ..., layer_0's Transpose; at last Aten(embedding). Then you would see all weights got transposed before the node that need immediate execution get deferred. 

Using the max_distance together with Reversed DFS from forward outputs, when it goes to the nodes that taking layer_11's Transpose, and another node N (generated by layer_10) as inputs, node N will be preferred to be output first because its distance to graph input is bigger than layer_11's Transpose.

Comment (AdamLouly) [reply onnxruntime/core/graph/graph.cc:1961]: Make sense, thank you!

**Review by AdamLouly — APPROVED** [review]
LGTM

**Comment by pengwa** [comment]
Thanks @adamlouly !!
